lib/subjects/remover.rb

```diff
@@ -0,0 +1,62 @@
+module Subjects
+  class Remover
+    class NonOrphan < StandardError; end
+
+    attr_reader :subject_id
+
+    def initialize(subject_id)
+      @subject_id = subject_id
+    end
+
+    def cleanup
+      if can_be_removed?
+        locations = orphan.locations
+        set_member_subjects = orphan.set_member_subjects
+        workflow_ids = orphan.workflows.map(&:id)
+        ActiveRecord::Base.transaction do
+          orphan.delete
+          locations.map(&:destroy)
+          set_member_subjects.map(&:destroy)
+        end
+        notify_cellect(workflow_ids)
+      else
+        raise_non_orphan_error
+      end
+    end
+
+    private
+
+    def can_be_removed?
+      !!orphan && no_talk_discussions?
+    end
+
+    def orphan
+     @orphan ||=
+       Subject
+       .where(id: subject_id)
+       .joins("LEFT OUTER JOIN classification_subjects ON classification_subjects.subject_id = subjects.id")
+       .where("subjects.id IS NOT NULL AND classification_subjects.subject_id IS NULL")
```

Comment:
Yeah - this can get rewritten, it was vestigial from a larger scope that I iterated over.